### PR TITLE
Support default values for case class fields

### PR DESCRIPTION
**Pull request by soujiro32167**
This is basically https://github.com/fd4s/vulcan/pull/566 with more comprehensive tests

### CODE REVIEW

**Comment by ayoub-benali** [comment]
Thanks for the PR. The CI is failing for a formatting issue, but locally AvroFieldDefaultSpec is failing also

**Comment by soujiro32167** [comment]
Ran `scalafmtAll` - test pass fine for me... 🤔 
Which test failures are you getting?

**Comment by ayoub-benali** [comment]
> Which test failures are you getting?

sbt clean did the trick. The tests pass now :)

File: modules/generic/src/test/scala/vulcan/generic/AvroFieldDefaultSpec.scala [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+final class AvroFieldDefaultSpec extends CodecBase {
+  describe("AvroFieldDefault") {
+    it("should create a schema with a default for a field") {
+      assert(Foo.codec.schema.exists(_.getField("a").defaultVal() == 1))
+      assert(Foo.codec.schema.exists(_.getField("b").defaultVal() == "foo"))
+    }
+
+    it("should fail when annotating an Option") {
+      assertSchemaError[InvalidDefault2]
```
Comment (ayoub-benali) [inline modules/generic/src/test/scala/vulcan/generic/AvroFieldDefaultSpec.scala:21]: It isn't clear for me why the schema derivation is supposed to fail here with 
`Left(AvroError(org.apache.avro.AvroTypeException: Invalid default for field a: "foo" not a ["null","string"]))`

The value `"foo"` has the right type, isn't it ?

Comment (soujiro32167) [reply modules/generic/src/test/scala/vulcan/generic/AvroFieldDefaultSpec.scala:21]: In [avro spec](https://avro.apache.org/docs/1.11.1/specification/#complex-types), it turns out that the order of members of a union determine the type of default you can have (I know, 😱 )

Specifically, the default can only correspond to the **first value**.

In `[null, string]`, the only `null` or `None` is a valid default

Comment (soujiro32167) [reply modules/generic/src/test/scala/vulcan/generic/AvroFieldDefaultSpec.scala:21]: Avro4s has a neat way of addressing this - its possible to determine the order of the elements in the schema using annotations. 

In vulcan, `Option` is always encoded `null` first, so its impossible to have a `Some("foo")` default at all

Comment (soujiro32167) [reply modules/generic/src/test/scala/vulcan/generic/AvroFieldDefaultSpec.scala:21]: Sadly this is only discoverable at test time right now.

**Comment by soujiro32167** [comment]
@ayoub-benali waiting for you to approve ci

File: modules/generic/src/test/scala/vulcan/generic/AvroFieldDefaultSpec.scala (hunk 1 above)
Comment (ayoub-benali) [reply modules/generic/src/test/scala/vulcan/generic/AvroFieldDefaultSpec.scala:21]: Oh I see, thanks for the explanation

**Review by ayoub-benali — COMMENTED** [review]
LGTM ! 
I will give a chance to @vlovgr or @bplommer for giving input, otherwise we can merge :)

**Review by vlovgr — APPROVED** [review]

**Comment by soujiro32167** [comment]
👏 👏 👏 Thank you!

**Pull request merged**